Lessons 1-6

alexey-dev0 opened the pull request:
13.09.2016
Lesson3

alexey-dev0 committed:
07.09.2016

alexey-dev0 committed:
08.09.2016

alexey-dev0 committed:
08.09.2016
Не сделал функцию circleInside.

alexey-dev0 committed:
09.09.2016
 Сделал функцию circleInside. Действительно, на бумаге всё стало очевидно.

alexey-dev0 committed:
13.09.2016

alexey-dev0 committed:
13.09.2016

alexey-dev0 committed:
Update List.kt

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +33,22 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
-fun ageDescription(age: Int): String = TODO()
+fun ageDescription(age: Int): String
+{
+    var Str = "$age"
```
~~Переменные принято называть с маленькой буквы.~~
Не относится к текущему Pull Request'у, но всё же

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +170,39 @@ fun sin(x: Double, eps: Double): Double = TODO()
+            count++
+            num/=10
+        }
+        num = n%10 * Math.pow(10.0, count.toDouble()).toInt()
```
Старайтесь избегать операций с числами с плавающей точкой в целочисленных задачах.
В Java и Kotlin нет функции для целочисленного возведения в степень, но вы можете произвести эту операцию в вашем цикле `while` вместо подсчёта количества цифр.

gagarski replied:
Всё ещё актуально.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +211,37 @@ fun revert(n: Int): Int = TODO()
+            number /= 10
+            pow++
+        }
+        return number == numLast && isPalindrome( (n-number*Math.pow(10.0, pow.toDouble()).toInt())/10 )
```
То же замечание про числа с плавающей точкой, что и в функции `revert()`.

gagarski replied:
актуально

gagarski commented:
recheck all

ice-phoenix commented:
recheck all

ice-phoenix commented:
recheck all

gagarski commented:
recheck all

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +182,15 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
+    if (list.size < 2) return list
+    for (i in list.size - 1 downTo 1) {
+        var sum = list[i]
+        for (j in 0..i - 1) sum += list[j]
```
На каждой итерации вы пересчитываете сумму начала списка. Сделайте так, чтобы время решения линейно (а не квадратично) зависело от длины списка.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +199,38 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
  * Результат разложения вернуть в виде строки, например 75 -> 3*5*5
  */
-fun factorizeToString(n: Int): String = TODO()
+fun factorizeToString(n: Int): String {
```
~~Данная задача может быть решена с использованием реализованной вами функции `factorize`.~~

gagarski replied:
done

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +258,19 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
  */
-fun convertToString(n: Int, base: Int): String = TODO()
+fun convertToString(n: Int, base: Int): String {
+    var string = ""
```
Данная задача может быть решена с использованием реализованной вами функции `convert`.

gagarski replied:
done

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +296,16 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
  */
-fun decimalFromString(str: String, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var sum = 0
```
Данная задача может быть решена с использованием реализованной вами функции `decimal`.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +340,52 @@ fun roman(n: Int): String = TODO()
-fun russian(n: Int): String = TODO()
\ No newline at end of file
+fun russian(n: Int): String {
+    val listStr1 = listOf("сто", "двести", "триста", "четыреста", "пятьсот", "шестьсот", "семьсот", "восемьсот", "девятьсот")
```
- Переменным лучше давать говорящие имена.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +340,52 @@ fun roman(n: Int): String = TODO()
+    val listStr2 = listOf("десять", "двадцать", "тридцать", "сорок", "пятьдесят", "шестьдесят", "семьдесят", "восемьдесят", "девяносто")
+    val listStr3 = listOf("одиннадцать", "двенадцать", "тринадцать", "четырнадцать", "пятнадцать", "шестнадцать", "семнадцать", "восемнадцать", "девятнадцать")
+    val listStr4 = listOf("один", "два", "три", "четыре", "пять", "шесть", "семь", "восемь", "девять")
+    val Thousands = listOf("тысяча", "тысячи", "тысяч")
```
Переменные принято называть с маленькой буквы.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -214,7 +315,23 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
-fun roman(n: Int): String = TODO()
+fun roman(n: Int): String {
+    var number = n
+    val listDigs = listOf(1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000)
```
Почитайте про ассоциативные массивы (`Map`) и функции `mapOf()` и `linkedMapOf()`

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +340,52 @@ fun roman(n: Int): String = TODO()
+    else if (number > 0) string += listStr4[number % 10 - 1]
+
+    if (string.length > 0 && string[string.length - 1] == ' ') string = string.substring(0, string.length - 1)
+    if (string.length > 0 && string[0] == ' ') string = string.substring(1, string.length)
```
Для этого у строки есть метод `trim()`

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -223,4 +340,52 @@ fun roman(n: Int): String = TODO()
+
+    if (number >= 100) {
+        string += listStr1[number / 100 % 10 - 1]
+        number -= number / 100 * 100
```
Подумайте, какую базовую арифметическую операцию вы реализуете этой строкой.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    val min = Math.min(m, n)
+    var num = min
+    var count = 1
+    while (num%max != 0){
```
- Может ли НОК быть меньше чем максимальное число?
- Почитайте про стандартные алгоритмы нахождения НОД/НОК

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +122,12 @@ fun maxDivisor(n: Int): Int = TODO()
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    for (i in 2..Math.min(m, n)){
```
Подумайте про связь данной задачи с предыдущими и про свойства взаимно простых чисел.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +149,19 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
  */
-fun sin(x: Double, eps: Double): Double = TODO()
+fun sin(x: Double, eps: Double): Double {
+    var result = x
```
Протестируйте вашу реализацию синуса на больших числах (в районе десятков или сотен PI). Правильно ли ваше решение? Попробуйте понять почему и придумать, как это исправить.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +170,39 @@ fun sin(x: Double, eps: Double): Double = TODO()
  */
-fun cos(x: Double, eps: Double): Double = TODO()
+fun cos(x: Double, eps: Double): Double {
+    var result = 1.0
```
См. коммент к синусу.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * Ряд Фибоначчи определён следующим образом: fib(1) = 1, fib(2) = 1, fib(n+2) = fib(n) + fib(n+1)
  */
-fun fib(n: Int): Int = TODO()
+fun fib(n: Int): Int = if (n < 3) 1 else fib(n-2) + fib(n-1)
```
Рекурсивное решение напрашивается, исходя из определения чисел Фибоначчи, но попробуйте посчитать, к скольки вызовам функции `fib()` приведёт вызов `fib(10)`, `fib(40)`, `fib(55)`, сколько среди них вызовов с одинаковыми аргументами и решить задачу более оптимальным способом.

gagarski commented:
Постарайтесь исправить текущие замечания и ошибки прежде чем приступать к следующим урокам.

alexey-dev0 committed:
18.09.2016
Fixed something

alexey-dev0 committed:
Merge remote-tracking branch 'origin/master'

alexey-dev0 committed:
18.09.2016
Fixed something again

alexey-dev0 committed:
18.09.2016
Some fixes

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +181,45 @@ fun sin(x: Double, eps: Double): Double = TODO()
-fun cos(x: Double, eps: Double): Double = TODO()
+fun cos(x: Double, eps: Double): Double {
+    var value = x
+    while (Math.abs(value) > 2*Math.PI) {
```
В Java и Kotlin есть операция взятия остатка от деления для Double.

alexey-dev0 replied:
Не знал, спасибо. Сделал с её помощью

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +154,25 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
-fun sin(x: Double, eps: Double): Double = TODO()
+fun sin(x: Double, eps: Double): Double {
+    var value = x
+    while (Math.abs(value) > 2*Math.PI) {
```
В Java и Kotlin есть операция взятия остатка от деления для Double.

alexey-dev0 committed:
19.09.2016
lesson3task1
Some fixes

alexey-dev0 committed:
19.09.2016
Lesson3Task1
Fixed cos & sin

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -1,6 +1,8 @@
 @file:Suppress("UNUSED_PARAMETER")
 package lesson3.task1
 
+import com.sun.org.apache.xpath.internal.operations.Bool
```
Избегайте ненужных импортов.

alexey-dev0 committed:
19.09.2016
Lesson5Task1
Done

alexey-dev0 committed:
19.09.2016
Lesson5Task1
Some fixes

alexey-dev0 commented:
Почему в последней задаче 5го урока можно прибавлять и убавлять значение в несуществующей ячейке?
Inputs:
    cells -> 0
    commands -> - --+

alexey-dev0 committed:
19.09.2016
Lesson3Task1
Some fixes

gagarski commented:
recheck all

gagarski commented:
> Почему в последней задаче 5го урока можно прибавлять и убавлять значение в несуществующей ячейке?
> Inputs:
> cells -> 0
> commands -> - --+

Тесты исправлены, но и у вас тоже ошибка. Во-первых, в задании сказано, какое исключение нужно бросать в этом случае. Во-вторых, над пустой лентой можно запускать пустой список команд.

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
> [Hard] lesson5.task1/computeDeviceCells
> Expected: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
> Actual: null
> Inputs:
> cells -> 310
> commands ->
> Exception: java.lang.StringIndexOutOfBoundsException: String index out of range: 0

Тут ваш баг.

> [Hard] lesson5.task1/computeDeviceCells
> Expected: java.lang.IllegalStateException: -1 is out-of-bounds for [0,310)
> Actual: null
> Inputs:
> cells -> 310
> commands -> <}-]<+[<>+}-<]-[-}+- ++-++ [ <<[<[]<{+>+< --<---}-<+]<<<-<<>+<< +<<--+<- +-+<{--}- -+--<+-+----}<+-]-----+<-+]-{-+[-<++-+----+-<- --}[}++<}- +-+->><+>< +---><+<-+--++----[--->-<[---<-{[+-}+<
> Exception: java.lang.IllegalArgumentException

А тут наш.

gagarski commented:
recheck all

alexey-dev0 committed:
22.09.2016
Lesson5Task1
Fix empty commands string

alexey-dev0 commented:
[Hard] lesson5.task1/firstDuplicateIndex

```
Expected: 7
Actual: -1
Inputs:
    str -> 1a ^ ^ m d9 ? eW x.H F W t 6 4 7rI 9
Exception: null
```

А какое тут слово повторяется с 7 индекса?

alexey-dev0 commented:
[Hard] lesson5.task1/computeDeviceCells

```
Expected: java.lang.IllegalStateException: -1 is out-of-bounds for [0,310)
Actual: null
Inputs:
    cells -> 310
    commands -> ++--}+--<- ---<<{+
Exception: java.lang.IllegalArgumentException
```

"}" не имеет пары, кинуто исключение IllegalArgumentException, вроде всё как надо

gagarski commented:
> А какое тут слово повторяется с 7 индекса?

Тут дело было в повторяющихся пробелов. Скорее всего на входе было что-то вроде 

```
1a     ^   ^ m d9 ? eW x.H F W t 6 4 7rI 9
```

но ГитХаб убрал лишние пробелы. Поскольку по заданию повторяющихся пробелов быть не должно, исправлено.

gagarski commented:
recheck assignment

gagarski commented:
recheck assignment

gagarski commented:
> [Hard] lesson5.task1/computeDeviceCells
> Expected: java.lang.IllegalArgumentException: No pair for bracket
> Actual: null
> Inputs:
> cells -> 1
> commands -> -{-+{<-[>->>]]-{->-
> Exception: java.lang.IllegalStateException

А теперь у вас баг.

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +60,27 @@ fun main(args: Array<String>) {
  */
-fun dateStrToDigit(str: String): String = TODO()
+fun dateStrToDigit(str: String): String {
+    val mounths = listOf("января", "февраля", "марта", "апреля", "мая", "июня", "июля", "августа", "сентября", "октября", "ноября", "декабря")
```
`months`

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +60,27 @@ fun main(args: Array<String>) {
+    if (parts.size < 3) return ""
+
+    val day: Int
+    val mounth: Int
```
`month`

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +147,31 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    try {
+
+        for (i in jumps) {
+            if ((i in intChars || i in allowedChars) == false) return -1
```
Помимо оператора `in` есть оператор `!in`. Применив закон де Моргана можно упростить условие:

``` kotlin
if (i !in intChars && i !in allowedChars) return -1
```

gagarski replied:
Далее то же самое в нескольких местах

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +267,13 @@ fun convertToString(n: Int, base: Int): String = TODO()
+fun decimal(digits: List<Int>, base: Int): Int {
+    var sum = 0
+    for (i in 0..digits.size - 1) {
+        sum += digits[i] * Math.pow(base.toDouble(), (digits.size - i - 1).toDouble()).toInt()
```
Это задача с целыми числами, не следует в таких переходить к Double без необходимости. Реализуйте `pow` для целых чисел, либо вычисляйте `base ^ n` в вашем цикле.

gagarski replied:
В `isPalindrome` пока висит такое же замечание

gagarski reviewed (changes requested):
(не заметил ранее)

alexey-dev0 committed:
01.10.2016
Lesson4Task1
Исправил ненужные преобразования в целые числа и обратно

alexey-dev0 committed:
03.10.2016
Lesson6Task2
Ошибки в тестах с траекториями короля и слона, но не могу понять из-за чего.

alexey-dev0 committed:
03.10.2016
Lesson6Task2
Исправил проверки клеток на правильность

alexey-dev0 committed:
03.10.2016
Lesson6Task2
Исправил проверку у слона.

alexey-dev0 committed:
03.10.2016
Lesson6Task2
Исправил проверку у слона снова.

gagarski commented:
recheck all

gagarski commented:
Из старого:
Осталось замечание про `Double` в `isPalindrome`.
Также не проходят тесты для `computeDeviceCells`

alexey-dev0 committed:
03.10.2016
Lesson6Task2
Исправлена функция isPalindrome

gagarski commented on src/lesson6/task2/Chess.kt:
```diff
@@ -182,7 +284,41 @@ fun kingTrajectory(start: Square, end: Square): List<Square> = TODO()
+    allList.add(start)
+    prevList.add(start)
+
+    val signX = listOf(-2, -2, -1, -1, 1, 1, 2, 2)
```
В `signX` и `signY` хранятся возможные ходы коня. Связывать их только индексом в соседних списках - плохой стиль.
Лучше хранить их в одном списке:

``` kotlin
val moves = listOf(Pair(-2, -1), Pair(-2, 1) /* ... */)
```

gagarski commented on src/lesson6/task2/Chess.kt:
```diff
@@ -204,4 +340,55 @@ fun knightMoveNumber(start: Square, end: Square): Int = TODO()
+    allList.add(start)
+    prevList.add(start)
+
+    val signX = listOf(-2, -2, -1, -1, 1, 1, 2, 2)
```
`listOf(Pair()...)`

gagarski replied:
`listOf(Pair()...)`

gagarski commented on src/lesson6/task2/Chess.kt:
```diff
@@ -204,4 +340,55 @@ fun knightMoveNumber(start: Square, end: Square): Int = TODO()
+    }
+
+    var curSq = end
+    while (curSq != start) {
```
Сможете ли вы объяснить, почему такой подход работает (сначала сложить ходы в формате "из X в Y" в список, а потом пройтись по этим парам и составить путь). Почему в результате получается **кратчайший** путь?

alexey-dev0 committed:
05.10.2016
Lesson5Task1
Переделал computeDeviceCells

alexey-dev0 commented:
А строку команд нужно сразу проверять на правильность или в процессе выполнения команд?
И ещё вопрос, в тестах есть такой: assertEquals(listOf(IllegalArgumentException()), computeDeviceCells(1, "-{-+{<-[>->>]]-{->-"))
Если сразу не проверять на правильность строку команд, то разве не должно бросить исключение IllegalStateException() ? Так как '{' и '[' по условию воспринимаются только если под датчиком 0, или я не прав?
Преобразования значения под датчиком:
0  >  -1  >  -2  >  -1  >  IllegalStateException()

gagarski commented:
recheck all

gagarski commented:
> А строку команд нужно сразу проверять на правильность или в процессе выполнения команд?

Возможно в задании это сформулировано недостаточно чётко. Убрал генерацию программ с непарными скобками. Но у вас всё равно баг.

> [Hard] lesson5.task1/computeDeviceCells
> java.lang.IllegalArgumentException : null
> [Hard] lesson5.task1/computeDeviceCells
> Expected: java.lang.IllegalStateException: -1 is out-of-bounds for [0,649)
> Actual:
> Inputs:
> cells -> 649
> commands -> +<[-<]--+<<+>-<->++-+[+< <--<<+]<+[<+--+<+---+><<<+>-]+<{}-+<-{<>-<- >- <+-+<--<-[>-+<-+--+} +<-<----{-<+<-><+<+<-<}<- +->] <<-<-+[-+-]< >--{+ -++-----[<-<+ -<<-}><<+-<-<-++--<-<-<+< -]-+{<--+}
> Exception: java.lang.IllegalArgumentException

alexey-dev0 committed:
05.10.2016
Lesson5Task1
computeDeviceCells

alexey-dev0 committed:
05.10.2016
Lesson5Task1
computeDeviceCells

gagarski commented:
> [Hard] lesson5.task1/plusMinus
> Expected: java.lang.IllegalArgumentException: java.lang.NumberFormatException: For input string: "-"
> Actual: 0
> Inputs:
> - expression ->

Здесь неправильный результат при вызове `plusMinus(" -")`

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -187,4 +383,57 @@ fun fromRoman(roman: String): Int = TODO()
+
+    fun checkOp(iter: Int, op: Char): Int {
+        val sign: Int
+        if (opList.indexOf(op) % 2 == 0) sign = -1 else sign = 1
```
`opList.indexOf(op) % 2 == 0` - излишне запутанный способ определения открывающей скобки. Лучше просто `op =='[' || op == '{'`.

alexey-dev0 committed:
03.10.2016
Lesson6Task2
Some fixes

alexey-dev0 committed:
06.10.2016
Lesson5Task1
Fixes

alexey-dev0 committed:
06.10.2016
Lesson6Task2
Fixes

alexey-dev0 committed:
06.10.2016
Lesson6Task1
Почему не работает фунцкия crossPoint?

alexey-dev0 commented:
не работает фунцкия crossPoint

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson1Task1

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson6Task1

MoshkinAlexey commented:
recheck all

MoshkinAlexey commented:
recheck all

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 committed:
Lesson6Task1

alexey-dev0 commented:
Как решать последнюю задачу в 6-1?

Ищем самые удалённые точки. Из центра отрезка между этими точками ищем наиболее удалённую. Строим окружность по этим трём точкам.
Такое решение неправильное?

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -97,37 +130,88 @@ data class Line(val point: Point, val angle: Double) {
+fun lineBySegment(s: Segment): Line {
+    val xDist = s.end.x - s.begin.x
+    val yDist = s.end.y - s.begin.y
+    val angle = if (xDist == 0.0 && yDist == 0.0) 0.0
```
Нужно ли это? `Math.atan()` нормально работает с нулём и бесконечностью. Случай, когда `xDist == 0.0 && yDist == 0.0` является некорректным (можно построить бесконечное число прямых), такие тесты у нас не генерируются.

gagarski commented on src/lesson6/task2/Chess.kt:
```diff
@@ -74,7 +101,24 @@ fun rookMoveNumber(start: Square, end: Square): Int = TODO()
+
+    if (currentColumn != end.column) {
+        currentColumn = end.column
+        addSquare()
```
Ваш `addSquare` излишне запутывает довольно простую функцию. По сути вы имитируете передачу в неё аргументов через локальные переменные. Если три раза сделать `list.add()`, то ваша функция станет понятнее.

gagarski commented on src/lesson6/task2/Chess.kt:
```diff
@@ -119,7 +170,41 @@ fun bishopMoveNumber(start: Square, end: Square): Int = TODO()
+    if (start == end) return list
+
+    if (currentColumn - currentRow != end.column - end.row) {
+        var log = false
```
Почитайте [про метки циклов](https://kotlinlang.org/docs/reference/returns.html) и избавьтесь от переменной `log`.

gagarski commented on src/lesson6/task2/Chess.kt:
```diff
@@ -157,7 +247,27 @@ fun kingMoveNumber(start: Square, end: Square): Int = TODO()
+    while (currentColumn != end.column || currentRow != end.row) {
+        if (end.column != currentColumn) currentColumn += signX
+        if (end.row != currentRow) currentRow += signY
+        addSquare()
```
См. `rookTrajectory` про `addSquare()`

gagarski commented:
> Такое решение неправильное?

Ваше решение, во-вервых, выдаёт окружность большего радиуса, чем ожидаемый, во-вторых, ваша окружность содержит не все точки из списка (можете проверить).

alexey-dev0 committed:
Lesson6Task2

alexey-dev0 committed:
Lesson6Task1

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck assignment

gagarski commented:
recheck assignment

gagarski commented:
recheck all

mglukhikh commented:
recheck all

mglukhikh commented:
recheck all

mglukhikh commented:
recheck all

mglukhikh commented:
recheck all

gagarski commented:
Замечаний не осталось, но бот почему-то не хочет перепроверять этот PR. 
Предлагаю создать новый и `minContainingCircle` решать уже там.
